tests/test_merge.py

```diff
@@ -379,6 +381,332 @@ def test_merge(
+        ),
+        param(
+            (
+                DictConfig(content={"foo": MISSING}, element_type=Optional[str]),
+                {"foo": "qux"},
+            ),
+            {"foo": "qux"},
+            str,
+            True,
+            id="missing_optional_str",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": MISSING}, element_type=Optional[str]),
+                {"foo": None},
+            ),
+            {"foo": None},
+            str,
+            True,
+            id="missing_optional_str_none",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": MISSING}, element_type=Optional[str]),
+                {"foo": MISSING},
+            ),
+            {"foo": MISSING},
+            str,
+            True,
+            id="missing_optional_str_missing",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=User),
+                {"foo": User("Bond")},
+            ),
+            {"foo": User("Bond")},
+            User,
+            False,
+            id="user",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=User),
+                {"foo": None},
+            ),
+            raises(ValidationError, match="field 'foo' is not Optional"),
+            None,
+            None,
+            id="user_none",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=User),
+                {"foo": MISSING},
+            ),
+            {"foo": MISSING},
+            User,
+            False,
+            id="user_missing",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=Optional[User]),
+                {"foo": User("Bond")},
+            ),
+            {"foo": User("Bond")},
+            User,
+            True,
+            id="optional_user",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=Optional[User]),
+                {"foo": None},
+            ),
+            {"foo": None},
+            User,
+            True,
+            id="optional_user_none",
+        ),
+        param(
+            (
+                DictConfig(content={"foo": User("Bond")}, element_type=Optional[User]),
+                {"foo": MISSING},
+            ),
+            {"foo": User("Bond")},
+            User,
+            True,
+            id="optional_user_missing",
+        ),
+        param(
+            (DictConfig(content={}, element_type=User), {"foo": User("Bond")}),
+            {"foo": User("Bond")},
+            User,
+            False,
+            id="new_user",
+        ),
+        param(
+            (DictConfig(content={}, element_type=User), {"foo": None}),
+            raises(ValidationError, match="child 'foo' is not Optional"),
```

Comment:
Minor comment: may be worth looking at the full error message to make sure it makes sense (the excerpt being tested could be confusing)

Reply:
In https://github.com/omry/omegaconf/pull/749/commits/0d8196cea552b6f409be7ca20942d200147fbcf1 I've updated the tests to match the full error message (excluding the one with id="new_str_none", which is currently not passing).

Reply:
Thanks -- though personally I tend to prefer checking snippets to make tests more concise (leaving full checks to test_errors.py) -- but anyway the new version looks good to me.

This actually confirms that this error message ("child 'foo' is not Optional") is a bit confusing (IMO). I can see where it may be coming from (we first add a `foo` node of type `User` then try to set it to `None`), but a clearer error could be something like: "Can't merge None values into a DictConfig whose element type is not Optional".

Definitely not something to change in this PR (and may not be worth changing it at all) -- I just thought I'd mention it because it caught my eye :)